components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new EntitlementException("Error while updating subscriber details", e);
+        } finally {
+            IdentityDatabaseUtil.closeConnection(connection);
+        }
+    }
+
+    /**
+     * Sets the base64 encoded secret value of the secret subscriber properties, if it has been updated.
+     *
+     * @param holder    publisher data holder
+     * @param oldHolder existing publisher data holder
+     */
+    private void populateProperties(PublisherDataHolder holder, PublisherDataHolder oldHolder) {
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+
+        for (PublisherPropertyDTO dto : propertyDTOs) {
+            if (StringUtils.isNotBlank(dto.getId()) && StringUtils.isNotBlank(dto.getValue()) && dto.isSecret()) {
+                PublisherPropertyDTO propertyDTO = null;
+                if (oldHolder != null) {
+                    propertyDTO = oldHolder.getPropertyDTO(dto.getId());
+                }
+                if (propertyDTO == null || !propertyDTO.getValue().equalsIgnoreCase(dto.getValue())) {
+                    try {
+                        String encryptedValue =
+                                CryptoUtil.getDefaultCryptoUtil().encryptAndBase64Encode(dto.getValue().getBytes());
+                        dto.setValue(encryptedValue);
+                    } catch (CryptoException e) {
+                        LOG.error("Error while encrypting secret value of subscriber. Secret would not be persisted.",
+                                e);
+                    }
+                }
+            }
+        }
+    }
+
+    private PublisherDataHolder getPublisherDataHolder(ResultSet resultSet, boolean returnSecrets) throws SQLException {
+
+        List<PublisherPropertyDTO> propertyDTOList = new ArrayList<>();
+        String moduleName;
+        do {
+            PublisherPropertyDTO dto = new PublisherPropertyDTO();
+
+            dto.setId(resultSet.getString(PROPERTY_ID));
+            dto.setValue(resultSet.getString(PROPERTY_VALUE));
+            dto.setDisplayName(resultSet.getString(DISPLAY_NAME));
+            dto.setDisplayOrder(resultSet.getInt(DISPLAY_ORDER));
+            dto.setRequired(resultSet.getBoolean(IS_REQUIRED));
+            dto.setSecret(resultSet.getBoolean(IS_SECRET));
+            dto.setModule(resultSet.getString(MODULE));
+
+            if (dto.isSecret() && returnSecrets) {
+                String password = dto.getValue();
+                try {
+                    password = new String(CryptoUtil.getDefaultCryptoUtil().base64DecodeAndDecrypt(dto.getValue()));
+                } catch (CryptoException e) {
+                    LOG.error("Error while decrypting secret value of subscriber.", e);
+                }
+                dto.setValue(password);
+            }
+
+            moduleName = resultSet.getString(ENTITLEMENT_MODULE_NAME);
```

Comment:
Wouldn't it be better to query the module name from a separate query?

If we do so, `getPublisherDataHolder()` method will be renamed to something like `populatePublisherProperties()` and building the PublisherDataHolder object moved to `getSubscriber()` itself..

Reply:
> Wouldn't it be better to query the module name from a separate query?

Coz despite we initialize it many iterations, its not a part of the propertyDTO, rather its used as the module name of all the properties..

Reply:
didn't go through with this considering having an extra db call for retrieving a single value vs having an extra column with the same values. an extra db call is better to avoid IMO